src/opnsense/service/templates/OPNsense/Proxy/wpad.dat

```diff
@@ -4,24 +4,20 @@
 */
 function FindProxyForURL(url, host) {
 {% if helpers.exists('OPNsense.proxy.pac.rule') %}
-{# define only if needed as because of performance issues #}
-{% set dl = '' %}
-{% set dstip = '' %}
-{% set is_resolvable = '' %}
-{% for match in helpers.toList('OPNsense.Proxy.pac.match') %}
-{%   if match.match_type == 'dns_domain_levels' %}
-{%     set dl = 'var dl =  dnsDomainLevels(host);' %}
+{# define only if needed because of performance issues #}
+{% set data = [] %}
+{% for match in helpers.toList('OPNsense.proxy.pac.match') %}
+{%   if 'var dl =  dnsDomainLevels(host);' not in data and match.match_type == 'dns_domain_levels' %}
```

Comment:
why - this needs to be unique in the array?

Reply:
the model should enforce constraints if it matters, a full string compare is quite sensitive / prone to errors. 
Worst case scenario, if the string duplicates, the browser will very likely still understand.

Reply:
@AdSchellevis this is no model data, but code that may cause performance issues and for that reason it is here to be so the result is cached between different matches. It should:

1. executed only once
2. added only when needed

In the final script this will render to 0-3 variable assignments.

Reply:
no model? `OPNsense.proxy.pac.match`

If you still want to deduplicate, it's better to move back to dictionaries.... or add some other smarter match, string matching like this is prone to errors.

Reply:
ok, I'll go back to the Hash - it is a 1:n join between something hardcoded in the template and the dynamic model on the other side.